### ISSUE TITLE
[Help Needed] [DataParallel/DataLoader/AMP] Severe slowdown on large dataset with fp16 precision and DataParallel wrapper

### ISSUE THREAD

**Issue by asigalov61** [pytorch/pytorch#114437]
### 🐛 Describe the bug

Hello PyTorch team! 

First of all, thank you very much for your great work on PyTorch 2. Its amazing! Please keep it up!

I am writing because I need help with a strange problem/possible bug...

When I use the following code with a large amount of training data, I am experiencing severe training speed loss/slowdown:
https://github.com/asigalov61/Giant-Music-Transformer/blob/main/Training-Code/Giant_Music_Transformer_Maker.ipynb

Everything works more or less ok on small amount of training data but after a certain threshold, the severe slowdown appears.

I troubleshooted the problem the best I could, and it seems to be related (most likely) to the DataParallel wrapper. I know the docs say to use DDP instead of DP, but I like and need DP because it works better for me and it is easier to use.

Torch version does not matter. I experience the same thing on all 2 versions, including the preview. Although preview (2.2) seem to work faster for some reason.

Also, if it is relevant, OS also does not matter. I experience the same problem on Ubuntu, Debian and Google Colab.

I would really appreciate if you can help me to solve this issue or at least point me out in the right direction so that I can do it myself.

Thank you.

Alex.

PS. I was not sure where to post it and I am not sure if it is a bug, so please assign the right category to this issue if it is wrong. And if you need more info, please ask.

### Versions

PyTorch version: 2.2.0.dev20231120+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.10 (default, May 26 2023, 14:05:08)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-73-generic-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA H100 PCIe
Nvidia driver version: 525.105.17
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 57 bits virtual
CPU(s):                          26
On-line CPU(s) list:             0-25
Thread(s) per core:              1
Core(s) per socket:              1
Socket(s):                       26
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           143
Model name:                      Intel(R) Xeon(R) Platinum 8480+
Stepping:                        8
CPU MHz:                         2000.000
BogoMIPS:                        4000.00
Virtualization:                  VT-x
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       832 KiB
L1i cache:                       832 KiB
L2 cache:                        104 MiB
L3 cache:                        416 MiB
NUMA node0 CPU(s):               0-25
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Unknown: No mitigations
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Mitigation; TSX disabled
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 wbnoinvd arat avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b fsrm md_clear serialize tsxldtrk avx512_fp16 arch_capabilities

Versions of relevant libraries:
[pip3] flake8==3.7.9
[pip3] numpy==1.23.5
[pip3] pytorch-triton==2.1.0+6e4932cda8
[pip3] torch==2.2.0.dev20231120+cu118
[pip3] torch-summary==1.4.5
[pip3] torchaudio==2.2.0.dev20231120+cu118
[pip3] torchvision==0.15.1
[pip3] triton==2.0.0
[conda] Could not collect

**Comment by mikaylagawarecki** [comment]
Hi, would it be possible to provide a minimal repro of the issue? Closing for now but feel free to re-open the issue if you have a minimal repro.

Otherwise we use github issues only for bug reports or feature requests. Please use the [forum](https://discuss.pytorch.org/) to ask questions

**Comment by asigalov61** [comment]
@mikaylagawarecki @malfet @soulitzer @scw 

Hey guys!

is there any update or info about this issue?

I still can't train in mixed precision mode. Everything works in fp32 and without DP. I use the latest pyTorch version: 2.4.1+cu124

Thank you so much.

Alex
